Relax fences for intrusive ptr's refcnt

mcfi opened the pull request:
Summary: Relax fences for intrusive ptr's refcnt dec op for performance testing.

lock needs acquire when the op succeeds and relaxed if the op is not. In addition, the expire call and the following refcnt reads were merged to remove one extra read.

incref does not need any fences because the caller should already have a valid reference. use_count follows the same reasoning.

decref only needs a release fence to make sure every write op prior to it has finished. When the refcnt goes to zero, there should be a acquire fence to make sure no read op reads stale data before the object is destructed. However, microbenchmark showed that the optimal fence for decref is not performing noticeably better than the current decref with acq-rel, so we keep decref as-is.

This change should have no material impact on x86, but for Arm64 (and other CPUs with weak memory models), it should boost performance.

ezyang commented:
Scott I will happily defer to you on this

mcfi commented on c10/util/intrusive_ptr.h:
```diff
@@ -208,14 +206,13 @@ inline uint32_t atomic_weakcount_increment(std::atomic<uint32_t>& weakcount) {
+// Both decrements need to be release for correctness.
 inline uint32_t atomic_refcount_decrement(std::atomic<uint32_t>& refcount) {
-  return refcount.fetch_sub(1, std::memory_order_acq_rel) - 1;
+  return refcount.fetch_sub(1, std::memory_order_release) - 1;
```
the decref op can be:
1. acq_rel, or
2. rel in cases which the refcnt is not 0, and rel followed by an acquire fence if refcnt is 0.
In both cases, when the refcnt hits 0, none of the resource release is reordered before the last writer to the object, which is needed for correctness..
My change implements the second (except that I used a seq_cst read followed by a release, which is used as an acquire fence), which should be less restrictive than libcpp's because the common case is when refcnt does not hit zero.

mcfi replied:
This requested change is moot because I reverted my change at that line.

mcfi commented on c10/util/intrusive_ptr.h:
```diff
@@ -208,14 +206,13 @@ inline uint32_t atomic_weakcount_increment(std::atomic<uint32_t>& weakcount) {
 
 inline uint32_t atomic_weakcount_decrement(std::atomic<uint32_t>& weakcount) {
-  return weakcount.fetch_sub(1, std::memory_order_acq_rel) - 1;
+  return weakcount.fetch_sub(1, std::memory_order_release) - 1;
```
Please see my comments above.

mcfi commented on c10/util/intrusive_ptr.h:
```diff
@@ -283,15 +280,21 @@ class intrusive_ptr final {
+      // an acquire fence.
       bool should_delete =
-          target_->weakcount_.load(std::memory_order_acquire) == 1;
+          target_->weakcount_.load(std::memory_order_seq_cst) == 1;
```
if the decref op is release, we can't use acquire here because the CPU might reorder the read before the released store. There should be an acquire fence before the read, we can combine the acquire fence with the read op in a seq_cst read, which is exactly what I'm doing here.

mcfi commented on c10/util/intrusive_ptr.h:
```diff
@@ -283,15 +280,21 @@ class intrusive_ptr final {
         const_cast<std::remove_const_t<TTarget>*>(target_)->release_resources();
         should_delete =
             detail::atomic_weakcount_decrement(target_->weakcount_) == 0;
+        if (should_delete) {
+          std::atomic_thread_fence(std::memory_order_acquire);
+        }
```
This acquire fence is needed for correctness since the decref only has the release semantics. Otherwise, target_ destruction might be reordered before the refcnt hits zero.

albanD commented on c10/util/intrusive_ptr.h:
```diff
@@ -196,10 +196,8 @@ TTarget* assign_ptr_(TTarget* rhs) {
-// unique() reliable.
 inline uint32_t atomic_refcount_increment(std::atomic<uint32_t>& refcount) {
-  return refcount.fetch_add(1, std::memory_order_acq_rel) + 1;
+  return refcount.fetch_add(1, std::memory_order_relaxed) + 1;
```
We do rely on use_count/unique being valid in things like the autograd engine. This can be used over multiple threads when considering worker threads, but I assume races are going to be unlikely. In the worst case there, we will do extra work (cloning instead of re-using the Tensor).

We also leverage unicity for the PyObject resurrection logic. Which is pretty cursed as well. But again, I guess unlikely to race this way.

swolchok replied:
@albanD, what makes race conditions hard to debug is that they are unlikely. At scale, unlikely things happen all the time. if unique() is load-bearing for correctness in the PyObject resurrection logic that makes me less inclined to do this.

WonderfulVoid replied:
Returning a value from `atomic_refcount_increment()` makes it a load. Changing the semantics from load-acquire to load-relaxed would affect any user which checks the return value and then depends on observing any related memory updates. What do we know about usage of this function?
If the return value isn't used, then better to remove it and the weakening from acquire to relaxed shouldn't matter.
Correspondingly, `atomic_refcount_increment()` is also a store with release semantics. We need to know if any user depends on that store-release. It could synchronize-with any other load-acquire from this refcount.
Ideally, the specification of this function should specify what memory ordering semantics it has. Perhaps we need a new function `atomic_refcount_increment_rlx()` which does not have any memory ordering semantics and use that in suitable places.

mcfi replied:
> Returning a value from `atomic_refcount_increment()` makes it a load. Changing the semantics from load-acquire to load-relaxed would affect any user which checks the return value and then depends on observing any related memory updates. What do we know about usage of this function? If the return value isn't used, then better to remove it and the weakening from acquire to relaxed shouldn't matter. Correspondingly, `atomic_refcount_increment()` is also a store with release semantics. We need to know if any user depends on that store-release. It could synchronize-with any other load-acquire from this refcount. Ideally, the specification of this function should specify what memory ordering semantics it has. Perhaps we need a new function `atomic_refcount_increment_rlx()` which does not have any memory ordering semantics and use that in suitable places.

Very good point. If you look at the file, atomic_refcount_increment returns to incref and retain_, neither of which returns the incremented value, so there is no user for the returned value. Same for atomic_weakcount_increment.

yfeldblum reviewed (approved):
From feedback left internally.
* Primary benefit of this PR is from relaxing the memory-order in increment, in addition to the memory-orders in use-count and lock.
* But splitting out a fence in decrement is likely a pessimization on modern hardware architectures. This part is reverted.

mcfi committed:
Relax fences for intrusive ptr's refcnt (#162072)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/162072

Relax fences for intrusive_ptr's refcnt ops.

lock needs acquire when the op succeeds and relaxed if the op is not. In addition, the expire call and the following refcnt reads were merged to remove one extra read.

incref does not need any fences because the caller should already have a valid reference. use_count follows the same reasoning.

decref only needs a release fence to make sure every write op prior to it has finished. When the refcnt goes to zero, there should be a acquire fence to make sure no read op reads stale data before the object is destructed. However, microbenchmark showed that the optimal fence for decref is not performing as well as the current decref with acq-rel, so we keep decref as-is.

This change should have no material impact on x86, but for Arm64 (and other CPUs with weak memory models), it should boost performance.

Test Plan:
All unit tests passed: buck run mode/opt //caffe2/c10/test:util_base_tests

Microbenchmark showed about ~12% performance improvement on Arm Grace servers:
buck run mode/opt //caffe2/c10/benchmark:intrusive_ptr
https://docs.google.com/spreadsheets/d/1XVu28qsYMShwh0-QdGjMOCYnS-mEQ0fSA96dYFF0IpQ/edit?usp=sharing
Raw data:
https://www.internalfb.com/phabricator/paste/view/P1934125420

https://www.internalfb.com/servicelab/experiment/5602137632/

Rollback Plan:

Reviewed By: yfeldblum

Differential Revision: D80648402

swolchok reviewed (approved):
current set of changes appears to be only "stop trying to make use_count/unique reliable", which is fine enough. I'm nervous about the autograd stuff but I've never seen it before and @albanD was OK with it so I'm not going to block here.

swolchok commented:
@pytorchbot merge
